File: algorithm/src/tree/FindDuplicateSubtrees.java
package tree;

// 652-寻找重复的子树
// find-duplicate-subtrees
//给定一棵二叉树，返回所有重复的子树。对于同一类的重复子树，你只需要返回其中任意一棵的根结点即可。
//
// 两棵树重复是指它们具有相同的结构以及相同的结点值。
//
// 示例 1：
//
//         1
//       / \
//      2   3
//     /   / \
//    4   2   4
//       /
//      4
//
//
// 下面是两个重复的子树：
//
//       2
//     /
//    4
//
//
// 和
//
//     4
//
//
// 因此，你需要以列表的形式返回上述重复子树的根结点。
// Related Topics 树 深度优先搜索 广度优先搜索 二叉树

import structure.TreeNode;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

/**
 * @author lihua
 * @since 2021/11/22
 */
public class FindDuplicateSubtrees {

    private List<TreeNode> resultList = new ArrayList<>();

    private HashMap<String, Integer> counter = new HashMap<>();

    public List<TreeNode> findDuplicateSubtrees(TreeNode root) {
        doFindDuplicateSubtrees(root);
        return resultList;
    }

    private String doFindDuplicateSubtrees(TreeNode p) {
        if (p == null) {
            return "#";
        }
        StringBuilder sb = new StringBuilder();
        String left = doFindDuplicateSubtrees(p.left);
        String right = doFindDuplicateSubtrees(p.right);
        // 这里用先序/后序都行，唯独不能用中序。因为中序序列不能唯一确定一棵树
        String tree = sb.append(left).append(",")
                .append(right).append(",")
                .append(p.val).toString();
        if (counter.containsKey(tree) && counter.get(tree) == 1) {
            resultList.add(p);
        }
        counter.put(tree, counter.getOrDefault(tree, 0) + 1);
        return tree;
    }
}

File: algorithm/go/tree/find-duplicate-subtrees.go
package tree

import (
	. "algorithm.com/structure"
	"strconv"
)

//给你一棵二叉树的根节点 root ，返回所有 重复的子树 。
//
// 对于同一类的重复子树，你只需要返回其中任意 一棵 的根结点即可。
//
// 如果两棵树具有 相同的结构 和 相同的结点值 ，则认为二者是 重复 的。
//
//
//
// 示例 1：
//
//
//
//
//输入：root = [1,2,3,4,null,2,4,null,null,4]
//输出：[[2,4],[4]]
//
// 示例 2：
//
//
//
//
//输入：root = [2,1,1]
//输出：[[1]]
//
// 示例 3：
//
//
//
//
//输入：root = [2,2,2,3,null,3,null]
//输出：[[2,3],[3]]
//
//
//
// 提示：
//
//
// 树中的结点数在 [1, 5000] 范围内。
// -200 <= Node.val <= 200
//
//
// Related Topics 树 深度优先搜索 哈希表 二叉树

func findDuplicateSubtrees(root *TreeNode) []*TreeNode {
	mapping := make(map[string]int, 5000)
	var res []*TreeNode
	var dfs func(node *TreeNode) string
	dfs = func(node *TreeNode) string {
		if node == nil {
			return "#"
		}
		curTree := dfs(node.Left) + "," +
			dfs(node.Right) + "," +
			strconv.Itoa(node.Val)
		if mapping[curTree] == 1 { // 重复子树只加入一次
			res = append(res, node)
		}
		mapping[curTree]++
		return curTree
	}
	dfs(root)
	return res
}
